### PR TITLE
Slice order test with bugfix

### PR DESCRIPTION
**Pull request by corinnebosley**
The problem that was occurring here was due to the difference in dimension indices before and after slicing.  For example, if I had a 4-dimensional cube and wanted to do cube.slices with the dim order as [3, 0, 1], upon slicing this order becomes [2, 0, 1].  This transition was not properly handled before; it had been attempted with the use of argsort(), but this was muddling things up and not sorting them properly.  

To fix the issue, I decided that after I had sliced the cube, I needed to remap the dimensions by converting them to their names and then back to their new indices to make sure that the order was correct.  To do this I have borrowed a piece of functionality that was used in cube.slices() which turns the dimension name into its index.

### CODE REVIEW

**Comment by corinnebosley** [comment]
@marqh @djkirkham

File: lib/iris/tests/unit/cube/test_Cube.py [hunk 1]
```diff
@@ -409,6 +409,160 @@ def test_kwargs(self):
         self.assertMaskedArrayEqual(expected_result, res_cube.data)
 
 
+class Testslices(tests.IrisTest):
```
Comment (djkirkham) [inline lib/iris/tests/unit/cube/test_Cube.py]: Can you change `Testslices` to `TestSlices` please?

File: lib/iris/tests/unit/cube/test_Cube.py [hunk 2]
```diff
@@ -409,6 +409,160 @@ def test_kwargs(self):
...
+    This class tests the slices capability of iris.cube.Cube
+    '''
+    def setUp(self):
+        '''
+        setup a 4D iris cube, each dimension is length 1.
+        The dimensions are;
+            dim1: height
+            dim2: time
+            dim3: longitude
+            dim4: latitude
```
Comment (djkirkham) [inline lib/iris/tests/unit/cube/test_Cube.py]: The standard order of dimensions in Iris is T, Z, Y ,X. You could probably use one of the cubes in `iris.tests.stock`

File: lib/iris/tests/unit/cube/test_Cube.py [hunk 3]
```diff
@@ -409,6 +409,160 @@ def test_kwargs(self):
...
+        '''
+        self.cube = iris.cube.Cube(np.array([[[[8.]]]]))
+        self.cube.add_dim_coord(iris.coords.DimCoord([0], "height"), [0])
+        self.cube.add_dim_coord(iris.coords.DimCoord([0], "time"), [1])
+        self.cube.add_dim_coord(iris.coords.DimCoord([0], "longitude"), [2])
+        self.cube.add_dim_coord(iris.coords.DimCoord([0], "latitude"), [3])
+
+    @staticmethod
+    def expectedcube_setup(dim1name, dim2name, dim3name):
```
Comment (djkirkham) [inline lib/iris/tests/unit/cube/test_Cube.py]: Please us snake case for function and variable names (separate each word with an underscore).

File: lib/iris/cube.py [hunk 4]
```diff
@@ -2663,6 +2663,33 @@ def _as_list_of_coords(self, names_or_coords):
...
+    def _dim_as_index(self, ref):
+        """
+        Convert coord name to index.
+        """
+        try:
+            coord = self._as_list_of_coords(ref)[0]
+            dim, = self.coord_dims(coord)
+            if dim not in self.coord_dims(coord):
+                msg = ('Requested an iterator over a coordinate ({}) '
+                       'which does not describe a dimension.')
```
Comment (djkirkham) [inline lib/iris/cube.py]: This error message doesn't seem to match the function

**Comment by corinnebosley** [comment]
@pelson I think this seems pretty useful so I'd like to keep this one, but give it to someone else to finish.  This makes me extremely uncomfortable but it's better than just throwing it away I reckon.

**Comment by pelson** [comment]
@corinnebosley - do you have a [SSCCE](sscce.org) for this one?

**Comment by corinnebosley** [comment]
@pelson a whatnow?

**Comment by dkillick** [comment]
> a whatnow?

@pelson seconded...

**Comment by djkirkham** [comment]
I'm guess he @pelson meant to link to here: http://sscce.org/

**Comment by pelson** [comment]
@corinnebosley - this isn't in a position where it can be merged at present. If still desired, could you please rebase, add a what's new, add a small/simple test, and provide a SSCCE so that a reviewer has some context of what this PR is seeking to resolve.

Please feel free to re-open when ready and ping me - I'll happily review 👍

**Pull request closed without merging**